atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/queue/ScalingSweepTaskSchedulerTest.java

```diff
@@ -0,0 +1,310 @@
+        schedulerEnabled.set(false);
+        scheduler.start(10);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 7 * INITIAL_PAUSE, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(8 * 10)).call();
+    }
+
+    @Test
+    public void whenManyEntriesAreSweptNewTaskSpawns() throws Exception {
+        when(sweepIteration.call()).thenReturn(
+                SUCCESS_LARGE,
+                SUCCESS_MEDIUM, SUCCESS_MEDIUM,
+                SUCCESS_LARGE, SUCCESS_MEDIUM);
+        scheduler.start(1);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(1 + 2 + 2 + 3 + 3 + 3)).call();
+    }
+
+    @Test
+    public void coolDownProtectsAgainstSpawningNewTasks() throws Exception {
+        ScalingSweepTaskScheduler schedulerWithCoolDown = createScheduler(Duration.ofDays(1));
+        when(sweepIteration.call()).thenReturn(SUCCESS_LARGE);
+        schedulerWithCoolDown.start(1);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(1 + 5)).call();
+    }
+
+    @Test
+    public void whenFewEntriesAreSweptTasksAreReduced() throws Exception {
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL, SUCCESS_MEDIUM);
+        scheduler.start(10);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(10 + 5 * 9)).call();
+    }
+
+    @Test
+    public void whenFewEntriesDoNotReduceToZeroTasks() throws Exception {
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL);
+        scheduler.start(1);
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(1 + 5)).call();
+    }
+
+    @Test
+    public void coolDownProtectsAgainstReducingTasks() throws Exception {
+        ScalingSweepTaskScheduler schedulerWithCoolDown = createScheduler(Duration.ofDays(1));
+        when(sweepIteration.call()).thenReturn(SUCCESS_SMALL);
+
+        schedulerWithCoolDown.start(2);
+
+        deterministicScheduler.tick(INITIAL_DELAY + 5 * DELAY, TimeUnit.MILLISECONDS);
+        verify(sweepIteration, times(2 + 5 * 2)).call();
+    }
+
+    @Test
+    public void attemptToIncreaseNumberOfThreadsPreventsReduction() throws Exception {
+        Duration coolDown = Duration.ofSeconds(1);
+        ScalingSweepTaskScheduler schedulerWithCoolDown = createScheduler(coolDown);
+        when(sweepIteration.call()).thenReturn(
+                SUCCESS_LARGE,
+                SUCCESS_LARGE,
+                SUCCESS_SMALL);
+        schedulerWithCoolDown.start(2);
+
+        Uninterruptibles.sleepUninterruptibly(coolDown.toMillis() / 2 + 1, TimeUnit.MILLISECONDS);
+        deterministicScheduler.tick(INITIAL_DELAY, TimeUnit.MILLISECONDS);
+
+        Uninterruptibles.sleepUninterruptibly(coolDown.toMillis() / 2 + 1, TimeUnit.MILLISECONDS);
```

Comment:
We should be able to avoid this: you can pass in a clock or rather `Supplier<Instant>` to the scheduler